docs/design/hierarchical-queue-on-capacity-plugin.md

```diff
@@ -0,0 +1,133 @@
+- `BuildVictimsPriorityQueue`: The current implementation of this function relies on `TaskOrderFn` and `JobOrderFn` for sorting victims. Two new functions, `VictimTaskOrderFn` and `VictimJobOrderFn`, will be introduced specifically for sorting victims in the context of hierarchical queues. 
+  
+  ```go
+  func (ssn *Session) BuildVictimsPriorityQueue(victims []*api.TaskInfo, preemptor *api.TaskInfo) *util.PriorityQueue {
+      victimsQueue := util.NewPriorityQueue(func(l, r interface{}) bool {
+          lv := l.(*api.TaskInfo)
+          rv := r.(*api.TaskInfo)
+          if lv.Job == rv.Job {
+              return ssn.VictimTaskOrderFn(l, r, preemptor)
+          }
+
+          return ssn.VictimJobOrderFn(ssn.Jobs[lv.Job], ssn.Jobs[rv.Job], preemptor)
+      })
+      for _, victim := range victims {
+          victimsQueue.Push(victim)
+      }
+      return victimsQueue
+  }
+  ```
+
+  - `VictimTaskOrderFn`: Prioritize tasks that belong to the same parent queue as the preemptor. If necessary, it will consider other tasks from the bottom up in the queue hierarchy.
+  - `VictimJobOrderFn`: Similar to `VictimTaskOrderFn`, prioritize jobs that belong to the same parent queue as the preemptor. If necessary, it will consider other jobs from the bottom up in the queue hierarchy.
+
+- `updateParentQueue`: Update the resource status of parent queues whenever the `updateShare` function is executed. It will traverse the queue hierarchy from the bottom up and update the resource information of parent queues accordingly.
+
+**Note:**
+
+- If a queue is deleted, its child queues will be deleted first, followed by the deletion of the queue itself. Similarly, if a queue is closed, its child queues will be closed first, and then the queue itself will be closed.
+- When creating a queue, the following logic will be added to the ValidatingAdmissionWebhook:
+  - If the parent queue of the queue to be created is currently scheduling jobs/podgroups, the creation of the new queue will not be allowed.
+  - The capability of the parent queue of the queue to be created must be greater than or equal to the capability of the new queue, and the deserved/guarantee of the parent queue must be greater than or equal to the sum of all its child queues (including the queue to be created). If this condition is not satisfied, the creation of the new queue will not be allowed.
+- The above modifications are primarily applicable when `EnabledHierarchy` is set to true. If the capacity plugin does not require hierarchical queue management, the existing implementations of these functions will be retained.
+
+### Limitations:
+
+- The current design does not support scheduling jobs/podgroups to non-leaf queues. Only leaf queues can directly schedule and allocate resources to jobs/podgroups.
+- Design relevant vcctl commands, such as commands to obtain the child queues of a specific queue or commands to retrieve the entire hierarchical queue structure.
```

Comment:
This seems not a limitation, it should be implemented, we can put this into `Key function modifications` part, and if this feature's priority is relatively low, we can implement it in stages.